### DISCUSSION TITLE
Make a circuit return a value

### DISCUSSION THREAD

**Discussion by jeljelih** [Consensys/gnark#830] [Q&A] [answered]
In all gnark examples I have seen, the circuit returns error or nil. Is there a way to make the circuit returns a computed value? In my usecase, I would like the circuit to compute the commitment of a value and returns the result.

**Comment by ivokub** [comment, marked as answer]
Hi @jeljelih - gnark circuits right now check the correctness of the computation given an input. But the inputs must be provided by the prover. Even though theoretically it would be possible to return the corresponding witness value from the circuit, it would require substantial architectural redesign. We do not have it on the roadmap and I'm not sure we are going to implement it in the future.

So right now, I would just recommend computing the commitment value natively and then use the value as witness to the prover and verifier.

**Reply by jeljelih** [reply]
Dear @ivokub thanks for your answer! I understand.
So I will compute the commitment out of the circuit and pass the value to the circuit.
To compute the MiMC hash function, I found this go module "github.com/consensys/gnark-crypto/ecc/bn254/fr/mimc". Is this the recommended module? 

Because I used it as follows:
```
mimc := mimc.NewMiMC() //load hash function
fmt.Println(mimc.Sum([]byte(strconv.Itoa(5000))))
```
Instead of returning the hash, it returned the input represented as a byte array.

**Reply by ivokub** [reply]
Yes, right now gnark-crypto outputs the hash value represented as a byte array. However, the byte array can be converted to field element by deserialization.

For example, have a look at [MiMC test](https://github.com/Consensys/gnark/blob/master/std/hash/mimc/mimc_test.go) to see that it is possible to assign bytes to circuit witness and it is converted to field element automatically by gnark.

There was another [ticket](https://github.com/Consensys/gnark/discussions/824) discussing the difference between the binary hasher and field hasher. Right now we make a distinction in gnark, but not in gnark-crypto. We may implement field hasher interface separately in gnark-crypto in the future.

**Reply by vmdamn** [reply]
@ivokub @jeljelih 

> To compute the MiMC hash function, I found this go module "github.com/consensys/gnark-crypto/ecc/bn254/fr/mimc". Is this the recommended module?

Yeah I think but somehow I am facing a problem.

> Yes, right now gnark-crypto outputs the hash value represented as a byte array. However, the byte array can be converted to field element by deserialization.

I tried to do Deserialization but in 

[https://github.com/Consensys/gnark-crypto/blob/master/ecc/bn254/fr/mimc/mimc.go](url)

the mimc.Sum() is appending hash to b, So it needs to be extracted and deserialized but it is not happening. Here is the code:

```
func mimcHash(data []byte) string {
 
  
    fetchIt := bn254.NewMiMC()
    hash := fetchIt.Sum(data)
    fmt.Println(hash)
    hashExtraction := hash[len(hash)-32:]
    fmt.Println("***mimcHash Function Data")
    fmt.Println(data)
    finalHash := hex.EncodeToString(hashExtraction)
    fmt.Println("***mimcHash Function Hash of Data")
    fmt.Println(finalHash)
    fmt.Println(reflect.TypeOf(finalHash))
    return finalHash

}

 ```
